keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1134 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This enhancement adds a `MutatingAdmissionPolicy` API to allow mutating
+admission control rules to be declared using CEL expressions. This continues the
+work started by the [`ValidatingAdmissionPolicy`
+API](/keps/sig-api-machinery/3488-cel-admission-control/README.md).
+
+Mutations can be declared in CEL by combining CEL's object instantiation
+and Server Side Apply's merge algorithms.
+
+## Motivation
+
+A large proportion of mutations perform relatively simple changes such as setting
+a label or annotation, or adding a sidecar container to a pod. Such mutations
+can be expressed conveniently in CEL, eliminating the need for the developmental
+and operational complexity of a webhook.
+
+Offering CEL based mutation also has other fundamental advantages over webhooks.
+CEL mutations can be declared in a way that allows the kube-apiserver to
+introspect the mutation and extract useful information about which fields the
+mutation reads and writes. This information can be useful when ordering
+mutations.
+
+### Goals
+
+- Provide an alternative to mutating webhooks for the vast majority of mutating
+  admission use cases.
+- Provide the in-tree extensions needed to build policy frameworks for
+  Kubernetes, again without requiring webhooks for the vast majority of use
+  cases.
+- Provide an out-of-tree implementation of this enhancement (using a webhook)
+  that is supported by the Kubernetes org to provide this enhancement
+  functionality to Kubernetes versions where this enhancement is not available.
+- Provide core functionality as a library so that use cases like GitOps,
+  CI/CD pipelines, and auditing can run the same CEL validation checks
+  that the API server does.
+
+### Non-Goals
```

Comment:
Question: will accessing user info (what is currently admission webhook's `user.Info` field) ever be possible?
For context, we use mutating webhook to set requester info on an object being created for a later authorization type of check https://github.com/cert-manager/cert-manager/blob/v1.12.0-beta.0/internal/plugin/admission/certificaterequest/identity/certificaterequest_identity.go#L65-L71

Reply:
Yes, it will. We provide access to info.  For ValidatingAdmissionPolicy we provided access to all the request info available to admission webhooks via a "request" variable. We will do the same here. I'll add a note in this KEP.

We also provide an "authorizer" variable that provides access to functions that can be used to perform authz checks.